ijmarshall/robotreviewer
# Feature/docker compose

## Problems

The existing Dockerfile includes several services installed at once and running in a single container. This *should* be improved using a docker-compose file in order to run *RobotReviewer* as a stack with separate services with the same standards of deployment that were applied to the other related services.

## Solution

The changes in this Pull Request include:
- RabbitMQ is initialized directly from the official Docker image.
- Grobid is initialized directly form the official Docker image.
- Separate Dockerfile and configuration files for running Bert-as-a-service server in a separate container.
- Replaced `run` script for `entrypoint.sh` allowing different commands from base RobotReviewer image.
- React app for Robotreviewer's Demo website is built in a preparation stage from an official Node Docker image in main Dockerfile.
- Removed following instructions from main Dockerfile: 
    * Node installation & build of static files for React app.
    * Java and Gradle installation and all Grobid related commands.
    * Bert installation and all Bert related commands.
    * Volumes and port exposure commands (moved to docker-compose).
    
- Deployment using docker-compose file with 6 services: 

    * **robotreviewer-web**: 
	    Flask App for Robotreviewer Demo website, running with Gunicorn WSGI. It also runs a cronjob for uploaded PDFs database cleanup every 12 hours.

    * **robotreviewer-api**: 
            Flask App for Robotreviewer API, running with Gunicorn WSGI. 
            
    * **robotreviewer-celery**: 
            Celery worker running annotating tasks involving Machine Learning.

    * **rabbitmq**:
            Messaging broker for Celery tasks.

    * **bert**:
            Bert as a service server.

    * **grobid**:            
            Grobid service for generating arcticles bibliografic data.
- Separate `server.py` and `server_api.py` scripts for Web & API Flask Apps for proper Celery App calls. 
- `config.json` file as a docker-compose `config` to be able to change the settings or modify RobotReviewer API keys without the need of rebuilding the image.
- Template file `.envTemplate` for necessary environment variables.
- Development compose file `docker-compose.dev.yml` runs **web** & **api** services  with Flask development server instead of Gunicorn WSGI.
- Compose file with GPU support `docker-compose.gpu.yml` runs **celery**, **bert** and **grobid** with access to GPU resources of the machine with Nvidia CUDA drivers.

## How to test
 
1. Follow the instructions at `README.md` for running RobotReviewer with Docker and test the demo website exposed at http://localhost:5050. Upload some PDFs and verify that the reports are generated correctly.
2. Using cURL or Postman requests with the configured API-Key, test RobotReviewer API endpoints at http://localhost:5051. 
3. (Optional) Run the update script for Pubmed articles from [Trialstreamer](https://github.com/ijmarshall/trialstreamer) while running RobotReviewer locally.
